include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -299,6 +311,29 @@ class Matrix
         std::shared_ptr<const Partition<local_index_type, global_index_type>>
             partition);
 
+    /**
+     * Reads a square matrix from the device_matrix_data structure and a global
+     * partition.
+     *
+     * The global size of the final matrix is inferred from the size of the
+     * partition. Both the number of rows and columns of the device_matrix_data
+     * are ignored.
+     *
+     * @note The matrix data can contain entries for rows other than those owned
+     *        by the process. Entries for those rows are discarded.
+     *
+     * @param data  The device_matrix_data structure.
+     * @param partition  The global row and column partition.
+     * @param assembly_mode  The mode of assembly.
+     *
+     * @return the index_map induced by the partitions and the matrix structure
+     */
+    void read_distributed(
+        const device_matrix_data<value_type, global_index_type>& data,
+        std::shared_ptr<const Partition<local_index_type, global_index_type>>
+            partition,
+        assembly_mode assembly_type);
```

Comment:
```suggestion
    void read_distributed(
        const device_matrix_data<value_type, global_index_type>& data,
        std::shared_ptr<const Partition<local_index_type, global_index_type>>
            partition,
        assembly_mode assembly_type = assmbly_mode::communicate);
```
you do not need to create double interface, but you will need to move `read_distributed(device_matrix_data, row_partition, col_partition)` out of public interface with new name or add explicit interface like you did but only for that. It is to avoid the ambiguous function call. all the other function will end up with this call